# Pandas


## Introduction to Pandas

### Introduction to Pandas

Pandas is an open-source library for real world data analysis in python. It is built on top of Numpy. Using Pandas, data can be cleaned, transformed, manipulated, and analyzed. It is suited for different kinds of data including tabular as in a SQL table or a Excel spreadsheets, time series data, observational or statistical datasets.

The steps involved to perform data analysis using Pandas are as follows:

<img src="Assests/11627398032146.PNG">


### Steps in data Analysis

#### Reading the data

The first step is to read the data. There are multiple formats in which data can be obtained such as '.csv', '.json', '.xlsx' etc.

Below are the examples:

<b>Example of an excel file:</b>

<img src="Assests/x641627962603383.PNG">

<b>Example of a json (javascript object notation) file:</b>

<img src="Assests/x611627962355796.PNG">

<b>Example of a csv (comma separated values) file:</b>

<img src="Assests/x631627962488063.PNG">


### Steps in data Analysis

#### Exploring the data

The next step is to explore the data. Exploring data helps to:

<ul>
    <li>know the shape(number of rows and columns) of the data</li>
    <li>understand the nature of the data by obtaining subsets of the data</li>
    <li>identify missing values and treat them accordingly</li>
    <li>get insights about the data using descriptive statistics</li> 
</ul>

#### Performing operations on the data

Some of the operations supported by pandas for data manipulation are as follows:

<ul>
    <li>Grouping operations</li> 
    <li>Sorting operations</li> 
    <li>Masking operations</li> 
    <li>Merging operations</li> 
    <li>Concatenating operations</li> 
</ul>

#### Visualizing data

The next step is to visualize the data to get a clear picture of various relationships among the data. The following plots can help visualize the data:

<ul>
    <li>Scatter plot</li>
    <li>Box plot</li>
    <li>Bar plot</li>
    <li>Histogram and many more</li>
</ul>

#### Generating Insights

All the above steps help generating insights about our data.


### Why Pandas

Pandas is one of the most popular data wrangling and analysis tools because it:

<ul>
    <li>has the capability to load huge sizes of data easily</li>
    <li>provides us with extremely streamlined forms of data representation</li>
    <li>can handle heterogenous data, has extensive set of data manipulation features and makes data flexible and customizable</li>
</ul>


## Introduction to Pandas Objects


### Getting started with Pandas

To get started with Pandas, Numpy and Pandas needs to be imported as shown below:


In [4]:
#Importing libraries
#python library for numerical and scientific computing. pandas is built on top of numpy
import numpy as np 
#importing pandas
import pandas as pd

In a nutshell, Pandas objects are advanced versions of NumPy structured arrays in which the rows and columns are identified with labels instead of simple integer indices.

The basic data structures of Pandas are Series and DataFrame.


### Pandas Series Object

Series is one dimensional labelled array. It supports different datatypes like integer, float, string etc. Let us understand more about series with the following example.

Consider the scenario where marks of students are given as shown in the following table:

<table>
    <tr>
        <th>Student ID</th>
        <th>Marks</th>
    </tr>
    <tr>
        <td>1</td>
        <td>78</td>
    </tr>
    <tr>
        <td>1</td>
        <td>92</td>
    </tr>
    <tr>
        <td>1</td>
        <td>36</td>
    </tr>
    <tr>
        <td>1</td>
        <td>64</td>  
    </tr>
    <tr>
        <td>5</td>
        <td>89</td>
    </tr>     
</table>

The pandas series object can be used to represent this data in a meaningful manner. Series is created using the following syntax:

<b>Syntax:</b>

<ul type="none">
<li><b>pd.Series(data, index, dtype)</b></li>
<li>data – It can be a list, a list of lists or even a dictionary.</li>
<li>index – The index can be explicitly defined for different valuesif required.</li>
<li>dtype – This represents the data type used in the series (optional parameter).</li>
</ul>


In [5]:
series = pd.Series(data = [78, 92, 36, 64, 89])  
series

0    78
1    92
2    36
3    64
4    89
dtype: int64

As shown in the above output, the series object provides the values along with their index attributes.

<b>Series.values</b> provides the values.


In [6]:
series.values

array([78, 92, 36, 64, 89], dtype=int64)

<b>Series.index</b> provides the index.


In [7]:
series.index

RangeIndex(start=0, stop=5, step=1)

#### Accessing data in series

Data can be accessed by the associated index using [ ].


In [8]:
series[1]

92

#### Slicing a series


In [9]:
series[1:3]

1    92
2    36
dtype: int64

### Custom Index in Series

By default, series creates an integer index. The custom index can also be defined.

For example, consider the following table containing car details:

<table>
    <tr>
        <th>Car Name</th>
        <th>Car Price</th>
    </tr>
    <tr>
        <td>Swift</td>
        <td>700000</td>
    </tr>
    <tr>
        <td>Jazz</td>
        <td>800000</td>
    </tr>
    <tr>
        <td>Civic</td>
        <td>1600000</td>
    </tr>
    <tr>
        <td>Altis</td>
        <td>1800000</td>
    </tr>
    <tr>
        <td>Gallardo</td>
        <td>30000000</td>
    </tr>
</table>
 
A Pandas series can be created using the following syntax:


In [10]:
data = pd.Series(data = [700000, 800000, 1600000, 1800000, 30000000], index = ['Swift', 'Jazz', 'Civic', 'Altis', 'Gallardo'])
data


Swift         700000
Jazz          800000
Civic        1600000
Altis        1800000
Gallardo    30000000
dtype: int64

Values can be accessed as:


In [11]:
data['Swift']

700000

In [12]:
data['Jazz': 'Gallardo']

Jazz          800000
Civic        1600000
Altis        1800000
Gallardo    30000000
dtype: int64

In this case, observations are that the output starts from Jazz and goes till Gallardo(inclusive). This is the fundamental difference between implicit and explicit indexing.


### Series as a specialized dictionary

Series can also be viewed as a specialized dictionary where the keys act as index and corresponding values act as values.

Let us create a series out of the dictionary data structure.


In [13]:
#Using dictionary to create a series
car_price_dict = {'Swift':  700000,
                       'Jazz' :  800000,
                       'Civic' : 1600000,
                       'Altis' : 1800000,
                       'Gallardo': 30000000
                      }
car_price = pd.Series(car_price_dict)
car_price

Swift         700000
Jazz          800000
Civic        1600000
Altis        1800000
Gallardo    30000000
dtype: int64

### Pandas DataFrame object

A series gives a useful way to view and manipulate one dimensional data. But when data is present in rows and columns, it becomes necessary to make use of the Pandas DataFrame object. A DataFrame is a collection of series where each series represents a column from a table.

For example, consider the following table containing car details:

<table>
    <tr>
        <th>Car Name</th>
        <th>Car Price</th>
        <th>Car Manufacturer</th>
    </tr>
    <tr>
        <td>Swift</td>
        <td>700000</td>
        <td>Maruti</td>
    </tr>
    <tr>
        <td>Jazz</td>
        <td>800000</td>
        <td>Honda</td>
    </tr>
    <tr>
        <td>Civic</td>
        <td>1600000</td>
        <td>Honda</td>
    </tr>
    <tr>
        <td>Altis</td>
        <td>1800000</td>
        <td>Toyota</td>
    </tr>
    <tr>
        <td>Gallardo</td>
        <td>30000000</td>
        <td>Lamborghini</td>
    </tr>
</table>


Let us create two series from two dictionaries - one containing car name and price and the other with car name and manufacturer.


In [14]:
#Creating a car price series with a dictionary
car_price_dict = {'Swift':  700000,
                       'Jazz' :  800000,
                       'Civic' : 1600000,
                       'Altis' : 1800000,
                       'Gallardo': 30000000
                      }
car_price = pd.Series(car_price_dict)
# Creating the car manufacturer series with a dictionary
car_man_dict = {'Swift' : 'Maruti',
                  'Jazz'   : 'Honda',
                  'Civic'  : 'Honda',
                  'Altis'  : 'Toyota',
                   'Gallardo' : 'Lamborghini'}
car_man = pd.Series(car_man_dict)
print(car_price)
print(car_man)

Swift         700000
Jazz          800000
Civic        1600000
Altis        1800000
Gallardo    30000000
dtype: int64
Swift            Maruti
Jazz              Honda
Civic             Honda
Altis            Toyota
Gallardo    Lamborghini
dtype: object


Let us create a Dataframe object using the series objects as shown below:

<b>Syntax:</b>

<ul >
    <li><b>pd.DataFrame(data, index, columns)</b></li>
    <li>data - data can contain Series or list-like objects. If data is a dictionary, column order follows insertion-order.</li>
    <li>index- index for dataframe that is created. By default, it will be RangeIndex(0, 1, 2, …, n) if no explicit index is provided</li>
    <li>columns-  If data contains column labels, it will use the same . Else, default to RangeIndex(0, 1, 2, …, n).</li>
</ul>


In [15]:
cars = pd.DataFrame({'Price': car_price , 'Manufacturer' : car_man})
cars

Price Manufacturer
Swift       700000       Maruti
Jazz        800000        Honda
Civic      1600000        Honda
Altis      1800000       Toyota
Gallardo  30000000  Lamborghini

The output shows the Dataframe containing multiple columns. The car names act as the indices and ‘Price’ and ‘Manufacturer’ act as the columns or 'features' of this small dataset.

To access individual features, the following code can be used:


In [16]:
cars['Price']

Swift         700000
Jazz          800000
Civic        1600000
Altis        1800000
Gallardo    30000000
Name: Price, dtype: int64

In [17]:
cars['Manufacturer']

Swift            Maruti
Jazz              Honda
Civic             Honda
Altis            Toyota
Gallardo    Lamborghini
Name: Manufacturer, dtype: object

### Ways to create a DataFrame

There are different approaches to create a DataFrame such as:

#### 1. From a single series object

A DataFrame is a collection of Series objects, and a single-column DataFrame can be constructed from a single Series:


In [18]:
#Using dictionary to create a series
car_price_dict = {'Swift':  700000,
                       'Jazz' :  800000,
                       'Civic' : 1600000,
                       'Altis' : 1800000,
                       'Gallardo': 30000000
                      }
car_price = pd.Series(car_price_dict)
car_price
#Creating a DataFrame from car_price Series
pd.DataFrame(car_price, columns=['Car Price'])

Car Price
Swift        700000
Jazz         800000
Civic       1600000
Altis       1800000
Gallardo   30000000

#### 2. From a list of dictionaries

Consider the following data of marks for four students.

<table>
    <tr>
        <th>Name</th>
        <th>Marks</th>
    </tr>
        <td>Subodh</td>
        <td>28</td>
    </tr>
    <tr>
        <td>Ram</td>
        <td>27</td>
    </tr>
    <tr>
        <td>Abdul</td>
        <td>26</td>
    </tr>
    <tr>
        <td>John</td>
        <td>28</td>
    </tr>
</table>

Following list of dictionaries can be used:


In [19]:
data = [{'Name': 'Subodh', 'Marks': 28},
        {'Name': 'Ram', 'Marks': 27}, 
        {'Name': 'Abdul', 'Marks': 26}, 
        {'Name': 'John', 'Marks': 28}]
pd.DataFrame(data)

Name  Marks
0  Subodh     28
1     Ram     27
2   Abdul     26
3    John     28

Suppose there is a following table to be represented as a dataframe ?

<table>
    <tr>
        <td>Subject</td>
        <td>Subodh</td>
        <td>Ram</td>
        <td>Abdul</td>
        <td>John</td>
    </tr>
    <tr>
        <td>Mathematics</td>
        <td>20</td>
        <td>25</td>
        <td>Not appeared</td>
        <td>Not appeared</td>
    </tr>
    <tr>
        <td>Physics</td>
        <td>Not appeared</td>
        <td>Not Appeared</td>
        <td>29</td>
        <td>24</td>
    </tr>
</table>


In [20]:
pd.DataFrame([{'Subodh':20, 'Ram':25},
              {'Abdul':29, 'John':24}], 
              index = ['Mathematics', 'Physics'])

Subodh   Ram  Abdul  John
Mathematics    20.0  25.0    NaN   NaN
Physics         NaN   NaN   29.0  24.0

Each dictionary element in the list is taken as a row . Index is representing different subjects.

Note: NaN(Not a Number) represents missing values.

#### 3. From a dictionary of series objects

A DataFrame can be constructed from a dictionary of Series objects:


In [21]:
#Using dictionary to create a series
car_price_dict = {'Swift':  700000,
                       'Jazz' :  800000,
                       'Civic' : 1600000,
                       'Altis' : 1800000,
                       'Gallardo': 30000000
                      }
car_price = pd.Series(car_price_dict)
car_man_dict = {'Swift' : 'Maruti',
                  'Jazz'   : 'Honda',
                  'Civic'  : 'Honda',
                  'Altis'  : 'Toyota',
                   'Gallardo' : 'Lamborghini'}
car_man = pd.Series(car_man_dict)
cars = pd.DataFrame({'Price': car_price , 'Manufacturer' : car_man})
cars

Price Manufacturer
Swift       700000       Maruti
Jazz        800000        Honda
Civic      1600000        Honda
Altis      1800000       Toyota
Gallardo  30000000  Lamborghini

#### 4. From an existing file

In most real world scenarios, the data is in different file formats like csv, xlsx, json etc. Pandas supports reading the data from these files. Below is an example of creating a DataFrame from a json file.


In [22]:
data_json = pd.read_json('Assests/data.json',)
data_json

Duration  Pulse  Maxpulse  Calories
0          60    110       130     409.1
1          60    117       145     479.0
2          60    103       135     340.0
3          45    109       175     282.4
4          45    117       148     406.0
..        ...    ...       ...       ...
164        60    105       140     290.8
165        60    110       145     300.4
166        60    115       145     310.2
167        75    120       150     320.4
168        75    125       150     330.4

[169 rows x 4 columns]

#### The axis keyword

One of the important parameters used while performing operations on DataFrames is 'axis'. Axis takes two values: 0 and 1.

axis = 0 represents row specific operations.

axis = 1 represents column specific operations.


## Working with Datasets

### Working with datasets in Pandas

#### Reading the data from XYZ custom cars

Pandas can read a variety of files. For example, a table of fixed width formatted lines (read_fwf), excel sheets (read_excel), html files (read_html), json files (read_json) etc.

<img src="Assests/51627400865828.PNG">

XYZ custom cars data is given in a csv format. This data is imported to a pandas DataFrame as shown below.

<b>Syntax:</b>
<ul type="none">
<li><b>pd.read_csv(filepath)</b></li>
<li>filepath - storage path of the file</li>
</ul>

In [23]:
import pandas as pd
import numpy as np
df = pd.read_csv('Assests/auto-mpg.csv')
df

mpg  cylinders  displacement  horsepower  weight  acceleration  \
0    18.0          8         307.0       130.0    3504          12.0   
1    15.0          8         350.0       165.0    3693          11.5   
2    18.0          8         318.0       150.0    3436          11.0   
3    16.0          8         304.0       150.0    3433          12.0   
4    17.0          8         302.0       140.0    3449          10.5   
..    ...        ...           ...         ...     ...           ...   
393  27.0          4         140.0        86.0    2790          15.6   
394  44.0          4          97.0        52.0    2130          24.6   
395  32.0          4         135.0        84.0    2295          11.6   
396  28.0          4         120.0        79.0    2625          18.6   
397  31.0          4         119.0        82.0    2720          19.4   

     model year  origin                   car name  
0            70       1  chevrolet chevelle malibu  
1            70       1          buick skylark 320  
2            70       1         plymouth satellite  
3            70       1              amc rebel sst  
4            70       1                ford torino  
..          ...     ...                        ...  
393          82       1            ford mustang gl  
394          82       2                  vw pickup  
395          82       1              dodge rampage  
396          82       1                ford ranger  
397          82       1                 chevy s-10  

[398 rows x 9 columns]

### Exploring the data

#### 1. Head and Tail

To view the first few rows or the last few rows, the functions that can be used are: df.head() and df.tail() respectively. If the number of rows to be viewed is not passed, then, the head and tail functions provides five rows by default.

In [24]:
df.head()

mpg  cylinders  displacement  horsepower  weight  acceleration  \
0  18.0          8         307.0       130.0    3504          12.0   
1  15.0          8         350.0       165.0    3693          11.5   
2  18.0          8         318.0       150.0    3436          11.0   
3  16.0          8         304.0       150.0    3433          12.0   
4  17.0          8         302.0       140.0    3449          10.5   

   model year  origin                   car name  
0          70       1  chevrolet chevelle malibu  
1          70       1          buick skylark 320  
2          70       1         plymouth satellite  
3          70       1              amc rebel sst  
4          70       1                ford torino

In [25]:
df.tail()

mpg  cylinders  displacement  horsepower  weight  acceleration  \
393  27.0          4         140.0        86.0    2790          15.6   
394  44.0          4          97.0        52.0    2130          24.6   
395  32.0          4         135.0        84.0    2295          11.6   
396  28.0          4         120.0        79.0    2625          18.6   
397  31.0          4         119.0        82.0    2720          19.4   

     model year  origin         car name  
393          82       1  ford mustang gl  
394          82       2        vw pickup  
395          82       1    dodge rampage  
396          82       1      ford ranger  
397          82       1       chevy s-10

#### 2. Describe

The describe function can be used to generate a quick summary of data statistics.

In [26]:
df.describe()

mpg   cylinders  displacement  horsepower       weight  \
count  398.000000  398.000000    398.000000  392.000000   398.000000   
mean    23.514573    5.454774    193.425879  104.469388  2970.424623   
std      7.815984    1.701004    104.269838   38.491160   846.841774   
min      9.000000    3.000000     68.000000   46.000000  1613.000000   
25%     17.500000    4.000000    104.250000   75.000000  2223.750000   
50%     23.000000    4.000000    148.500000   93.500000  2803.500000   
75%     29.000000    8.000000    262.000000  126.000000  3608.000000   
max     46.600000    8.000000    455.000000  230.000000  5140.000000   

       acceleration  model year      origin  
count    398.000000  398.000000  398.000000  
mean      15.568090   76.010050    1.572864  
std        2.757689    3.697627    0.802055  
min        8.000000   70.000000    1.000000  
25%       13.825000   73.000000    1.000000  
50%       15.500000   76.000000    1.000000  
75%       17.175000   79.000000    2.000000  
max       24.800000   82.000000    3.000000

#### 3. Info

To know about the datatypes and number of rows containing null values for respective columns, the info() function can be used.

In [27]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 398 entries, 0 to 397
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   mpg           398 non-null    float64
 1   cylinders     398 non-null    int64  
 2   displacement  398 non-null    float64
 3   horsepower    392 non-null    float64
 4   weight        398 non-null    int64  
 5   acceleration  398 non-null    float64
 6   model year    398 non-null    int64  
 7   origin        398 non-null    int64  
 8   car name      398 non-null    object 
dtypes: float64(4), int64(4), object(1)
memory usage: 28.1+ KB


#### 4. Dropping null values

It can be observed that the ‘horsepower’ attribute has some null values. The easiest approach is to remove the rows with any null values. This can be achieved using dropna() function.

In [28]:
df.dropna(inplace = True)
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 392 entries, 0 to 397
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   mpg           392 non-null    float64
 1   cylinders     392 non-null    int64  
 2   displacement  392 non-null    float64
 3   horsepower    392 non-null    float64
 4   weight        392 non-null    int64  
 5   acceleration  392 non-null    float64
 6   model year    392 non-null    int64  
 7   origin        392 non-null    int64  
 8   car name      392 non-null    object 
dtypes: float64(4), int64(4), object(1)
memory usage: 30.6+ KB


After dropping the rows with null horsepower values, it can be observed that the number of rows has been reduced to 392.

<b>Note: </b>
<ul>
<li>'inplace' makes changes to the original DataFrame.</li>
<li>df.fillna(condition) can be used to fill all the missing values. The missing values are filled with mean, median, mode, or constant values.</li>
</ul>

#### 5. Selecting a subset of the data

In addition to data access techniques, pandas also provides techniques for indexing and selection. Selecting a specific column in a DataFrame can be achieved in following ways: 
<ul>
<li>Passing the column name as shown below: </li>
</ul>

In [29]:
df['car name']

0      chevrolet chevelle malibu
1              buick skylark 320
2             plymouth satellite
3                  amc rebel sst
4                    ford torino
                 ...            
393              ford mustang gl
394                    vw pickup
395                dodge rampage
396                  ford ranger
397                   chevy s-10
Name: car name, Length: 392, dtype: object

Output is a Series containing car names.

 
<ul>
<li>Passing the column name as a list as shown below: </li>
</ul>

In [30]:
df[['car name']]

car name
0    chevrolet chevelle malibu
1            buick skylark 320
2           plymouth satellite
3                amc rebel sst
4                  ford torino
..                         ...
393            ford mustang gl
394                  vw pickup
395              dodge rampage
396                ford ranger
397                 chevy s-10

[392 rows x 1 columns]

Output is a DataFrame containing just one column.
<ul>
<li>To extract the subset of the data, we can pass the column names in a list as shown below:</li>
</ul>

In [31]:
df[['car name', 'origin', 'model year', 'mpg']]

car name  origin  model year   mpg
0    chevrolet chevelle malibu       1          70  18.0
1            buick skylark 320       1          70  15.0
2           plymouth satellite       1          70  18.0
3                amc rebel sst       1          70  16.0
4                  ford torino       1          70  17.0
..                         ...     ...         ...   ...
393            ford mustang gl       1          82  27.0
394                  vw pickup       2          82  44.0
395              dodge rampage       1          82  32.0
396                ford ranger       1          82  28.0
397                 chevy s-10       1          82  31.0

[392 rows x 4 columns]

### Indexers

#### Setting custom index:
Custom index can be set to the DataFrame according to the requirements. The following example depicts the same:

In [32]:
#creating a subset using head
df_head = df.head()
#Setting name as custom index
df_head.set_index('car name', inplace = True)
df_head

mpg  cylinders  displacement  horsepower  weight  \
car name                                                                       
chevrolet chevelle malibu  18.0          8         307.0       130.0    3504   
buick skylark 320          15.0          8         350.0       165.0    3693   
plymouth satellite         18.0          8         318.0       150.0    3436   
amc rebel sst              16.0          8         304.0       150.0    3433   
ford torino                17.0          8         302.0       140.0    3449   

                           acceleration  model year  origin  
car name                                                     
chevrolet chevelle malibu          12.0          70       1  
buick skylark 320                  11.5          70       1  
plymouth satellite                 11.0          70       1  
amc rebel sst                      12.0          70       1  
ford torino                        10.5          70       1

'iloc' and 'loc' are the two indexing techniques that help us in selecting specific rows and columns.

1. iloc- Access a group of rows and columns by integer index.

The ‘iloc’ indexer follows implicit index.  

    Syntax - df.iloc[Rows, Columns]

In the following demos, 'df' refers to XYZ Custom Cars DataFrame.

In [33]:
df

mpg  cylinders  displacement  horsepower  weight  acceleration  \
0    18.0          8         307.0       130.0    3504          12.0   
1    15.0          8         350.0       165.0    3693          11.5   
2    18.0          8         318.0       150.0    3436          11.0   
3    16.0          8         304.0       150.0    3433          12.0   
4    17.0          8         302.0       140.0    3449          10.5   
..    ...        ...           ...         ...     ...           ...   
393  27.0          4         140.0        86.0    2790          15.6   
394  44.0          4          97.0        52.0    2130          24.6   
395  32.0          4         135.0        84.0    2295          11.6   
396  28.0          4         120.0        79.0    2625          18.6   
397  31.0          4         119.0        82.0    2720          19.4   

     model year  origin                   car name  
0            70       1  chevrolet chevelle malibu  
1            70       1          buick skylark 320  
2            70       1         plymouth satellite  
3            70       1              amc rebel sst  
4            70       1                ford torino  
..          ...     ...                        ...  
393          82       1            ford mustang gl  
394          82       2                  vw pickup  
395          82       1              dodge rampage  
396          82       1                ford ranger  
397          82       1                 chevy s-10  

[392 rows x 9 columns]

In [34]:
df.iloc[2,1]

8

In [35]:
df.iloc[2,-1]

'plymouth satellite'

In [36]:
df.iloc[1:5, 4:6]

weight  acceleration
1    3693          11.5
2    3436          11.0
3    3433          12.0
4    3449          10.5

2. loc- Access a group of rows and columns by custom index.

The loc indexer follows explicit indexing.

In [37]:
#creating a subset using head. 'df' refers to XYZ Custom Cars DataFrame.
df_head = df.head()
#Setting name as custom index
df_head.set_index('car name', inplace = True)
df_head.loc['buick skylark 320': 'amc rebel sst']

mpg  cylinders  displacement  horsepower  weight  \
car name                                                                
buick skylark 320   15.0          8         350.0       165.0    3693   
plymouth satellite  18.0          8         318.0       150.0    3436   
amc rebel sst       16.0          8         304.0       150.0    3433   

                    acceleration  model year  origin  
car name                                              
buick skylark 320           11.5          70       1  
plymouth satellite          11.0          70       1  
amc rebel sst               12.0          70       1

To select a subset of columns, the column names can be passed as a list.

<b>Note:</b> While retrieving records using loc, the upper range of slice is inclusive.

In [38]:
#Subsetting from the full dataset
df.loc[0:5, ['cylinders', 'horsepower', 'car name']]

cylinders  horsepower                   car name
0          8       130.0  chevrolet chevelle malibu
1          8       165.0          buick skylark 320
2          8       150.0         plymouth satellite
3          8       150.0              amc rebel sst
4          8       140.0                ford torino
5          8       198.0           ford galaxie 500

### Adding/Removing columns in a DataFrame

Consider the following table:
<table>
    <tr>
        <th>Students</th>
        <th>Chemistry</th>
        <th>Physics</th>
        <th>Mathematics</th>
        <th>English</th>
    </tr>
    <tr>
        <td>Subodh</td>
        <td>67</td>
        <td>45</td>
        <td>50</td>
        <td>19</td>
    </tr>
    <tr>
        <td>Ram</td>
        <td>90</td>
        <td>92</td>
        <td>87</td>
        <td>90</td>
    </tr>
    <tr>
        <td>Abdul</td>
        <td>66</td>
        <td>72</td>
        <td>81</td>
        <td>72</td>
    </tr>
    <tr>
        <td>John</td>
        <td>32</td>
        <td>40</td>
        <td>12</td>
        <td>68</td>
    </tr>
</table>

The teacher wants to insert a ‘Total marks’ column which gives the sum of marks of all subjects.

In [39]:
marks = {'Chemistry': [67,90,66,32], 
        'Physics': [45,92,72,40],  
        'Mathematics': [50,87,81,12],  
        'English': [19,90,72,68]}
marks_df = pd.DataFrame(marks, index = ['Subodh', 'Ram', 'Abdul', 'John'])
marks_df

Chemistry  Physics  Mathematics  English
Subodh         67       45           50       19
Ram            90       92           87       90
Abdul          66       72           81       72
John           32       40           12       68

In [40]:
marks_df['Total'] = marks_df['Chemistry'] + marks_df['Physics'] + marks_df['Mathematics'] + marks_df['English']
marks_df

Chemistry  Physics  Mathematics  English  Total
Subodh         67       45           50       19    181
Ram            90       92           87       90    359
Abdul          66       72           81       72    291
John           32       40           12       68    152

To drop a feature:

In [41]:
marks_df.drop(columns = 'Total', inplace = True)

## Operations in Pandas

### Filtering operation

#### Problem statement : Retrieve details of all the cars built in year 72.

#### Solution:

In [42]:
df.loc[df['model year'] == 72 ].head()


mpg  cylinders  displacement  horsepower  weight  acceleration  \
57  24.0          4         113.0        95.0    2278          15.5   
58  25.0          4          97.5        80.0    2126          17.0   
59  23.0          4          97.0        54.0    2254          23.5   
60  20.0          4         140.0        90.0    2408          19.5   
61  21.0          4         122.0        86.0    2226          16.5   

    model year  origin               car name  
57          72       3  toyota corona hardtop  
58          72       1     dodge colt hardtop  
59          72       2      volkswagen type 3  
60          72       1         chevrolet vega  
61          72       1    ford pinto runabout

#### Problem statement : Retrieve details of all the cars built in Japan having 6 cylinders
#### Solution:

In [43]:
df.loc[(df['origin'] == 2) & (df['cylinders'] == 6)]

mpg  cylinders  displacement  horsepower  weight  acceleration  \
211  16.5          6         168.0       120.0    3820          16.7   
275  17.0          6         163.0       125.0    3140          13.6   
277  16.2          6         163.0       133.0    3410          15.8   
360  30.7          6         145.0        76.0    3160          19.6   

     model year  origin            car name  
211          76       2  mercedes-benz 280s  
275          78       2         volvo 264gl  
277          78       2       peugeot 604sl  
360          81       2        volvo diesel

#### Problem Statement:
XYZ Custom Cars want to categorize cars in different categories as follows:
<table>
    <tr>
        <th>Category</th>
        <th>Description</th>
        <th>Features coming in play</th>
    </tr>
    <tr>
        <td>Fuel efficient</td>
        <td>Cars designed with low power and high fuel efficiency</td>
        <td>High MPG, Low Horsepower, Low weight</td>
    </tr>
    <tr>
        <td>Muscle Cars</td>
        <td>Intermediate sized cars designed for high performance</td>
        <td>High displacement, High horsepower, Moderate weight</td>
    </tr>
    <tr>
        <td>SUV</td>
        <td>Big sized cars designed for high performance, long distance trips and family comfort</td>
        <td>High horsepower, High weight</td>
    </tr>
    <tr>
        <td>Racecar</td>
        <td>Cars specifically designed for race tracks</td>
        <td>Low weight, High acceleration</td>
    </tr>
</table>

Solution:
Their experienced engineers and mechanics have come up with the following parameters for these categories:
<table>
    <tr>
        <th>Category</th>
        <th>Description</th>
        <th>Features coming in play</th>
    </tr>
    <tr>
        <td>Fuel efficient</td>
        <td>Cars designed with low power and high fuel efficiency</td>
        <td>MPG > 29, Horsepower < 93.5, Weight < 2500</td>
    </tr>
    <tr>
        <td>Muscle Cars</td>
        <td>Intermediate sized cars designed for high performance</td>
        <td>Displacement > 262, Horsepower > 126, Weight in range[2800,3600]</td>
    </tr>
    <tr>
        <td>SUV</td>
        <td>Big sized cars designed for high performance, long distance trips and family comfort</td>
        <td>Horsepower > 140, Weight > 4500 </td>
    </tr>
    <tr>
        <td>Racecar</td>
        <td>Cars specifically designed for race tracks</td>
        <td>Weight < 2223, acceleration > 17 </td>
    </tr>
</table>

Let us see how we can find out the cars belonging to these categories based on the given parameters.

In [44]:
# Fuel efficient
# MPG > 29, Horsepower < 93.5,
# Weight < 2500
df.loc[(df['mpg'] > 29) & (df['horsepower'] < 93.5) & (df['weight'] < 2500)]

mpg  cylinders  displacement  horsepower  weight  acceleration  \
51   30.0          4          79.0        70.0    2074          19.5   
52   30.0          4          88.0        76.0    2065          14.5   
53   31.0          4          71.0        65.0    1773          19.0   
54   35.0          4          72.0        69.0    1613          18.0   
129  31.0          4          79.0        67.0    1950          19.0   
..    ...        ...           ...         ...     ...           ...   
384  32.0          4          91.0        67.0    1965          15.7   
385  38.0          4          91.0        67.0    1995          16.2   
391  36.0          4         135.0        84.0    2370          13.0   
394  44.0          4          97.0        52.0    2130          24.6   
395  32.0          4         135.0        84.0    2295          11.6   

     model year  origin             car name  
51           71       2          peugeot 304  
52           71       2            fiat 124b  
53           71       3  toyota corolla 1200  
54           71       3          datsun 1200  
129          74       3          datsun b210  
..          ...     ...                  ...  
384          82       3   honda civic (auto)  
385          82       3        datsun 310 gx  
391          82       1    dodge charger 2.2  
394          82       2            vw pickup  
395          82       1        dodge rampage  

[81 rows x 9 columns]

In [45]:
# Muscle cars
# Displacement >262, Horsepower > 126, Weight in range[2800, 3600]
df.loc[(df['displacement'] > 262) & (df['horsepower'] > 126) & (df['weight'] >=2800) & (df['weight'] <= 3600)]

mpg  cylinders  displacement  horsepower  weight  acceleration  \
0    18.0          8         307.0       130.0    3504          12.0   
2    18.0          8         318.0       150.0    3436          11.0   
3    16.0          8         304.0       150.0    3433          12.0   
4    17.0          8         302.0       140.0    3449          10.5   
10   15.0          8         383.0       170.0    3563          10.0   
13   14.0          8         455.0       225.0    3086          10.0   
121  15.0          8         318.0       150.0    3399          11.0   
166  13.0          8         302.0       129.0    3169          12.0   
251  20.2          8         302.0       139.0    3570          12.8   
262  19.2          8         305.0       145.0    3425          13.2   
264  18.1          8         302.0       139.0    3205          11.2   

     model year  origin                      car name  
0            70       1     chevrolet chevelle malibu  
2            70       1            plymouth satellite  
3            70       1                 amc rebel sst  
4            70       1                   ford torino  
10           70       1           dodge challenger se  
13           70       1       buick estate wagon (sw)  
121          73       1             dodge dart custom  
166          75       1               ford mustang ii  
251          78       1          mercury monarch ghia  
262          78       1  chevrolet monte carlo landau  
264          78       1                   ford futura

In [46]:
# SUV
# Horsepower > 140 , Weight > 4500
df.loc[(df['horsepower'] > 140) & (df['weight'] >=4500)]

mpg  cylinders  displacement  horsepower  weight  acceleration  \
25   10.0          8         360.0       215.0    4615          14.0   
28    9.0          8         304.0       193.0    4732          18.5   
42   12.0          8         383.0       180.0    4955          11.5   
43   13.0          8         400.0       170.0    4746          12.0   
44   13.0          8         400.0       175.0    5140          12.0   
67   11.0          8         429.0       208.0    4633          11.0   
68   13.0          8         350.0       155.0    4502          13.5   
90   12.0          8         429.0       198.0    4952          11.5   
94   13.0          8         440.0       215.0    4735          11.0   
95   12.0          8         455.0       225.0    4951          11.0   
103  11.0          8         400.0       150.0    4997          14.0   
104  12.0          8         400.0       167.0    4906          12.5   
105  13.0          8         360.0       170.0    4654          13.0   
137  13.0          8         350.0       150.0    4699          14.5   
156  16.0          8         400.0       170.0    4668          11.5   
159  14.0          8         351.0       148.0    4657          13.5   

     model year  origin                      car name  
25           70       1                     ford f250  
28           70       1                      hi 1200d  
42           71       1             dodge monaco (sw)  
43           71       1      ford country squire (sw)  
44           71       1           pontiac safari (sw)  
67           72       1               mercury marquis  
68           72       1          buick lesabre custom  
90           73       1      mercury marquis brougham  
94           73       1  chrysler new yorker brougham  
95           73       1      buick electra 225 custom  
103          73       1              chevrolet impala  
104          73       1                  ford country  
105          73       1        plymouth custom suburb  
137          74       1      buick century luxus (sw)  
156          75       1              pontiac catalina  
159          75       1                      ford ltd

In [47]:
# Racecar
# Weight <2223, acceleration > 17
df.loc[(df['acceleration'] > 17) & (df['weight'] < 2223)]

mpg  cylinders  displacement  horsepower  weight  acceleration  \
19   26.0          4          97.0        46.0    1835          20.5   
51   30.0          4          79.0        70.0    2074          19.5   
53   31.0          4          71.0        65.0    1773          19.0   
54   35.0          4          72.0        69.0    1613          18.0   
55   27.0          4          97.0        60.0    1834          19.0   
56   26.0          4          91.0        70.0    1955          20.5   
79   26.0          4          96.0        69.0    2189          18.0   
102  26.0          4          97.0        46.0    1950          21.0   
117  29.0          4          68.0        49.0    1867          19.5   
129  31.0          4          79.0        67.0    1950          19.0   
131  32.0          4          71.0        65.0    1836          21.0   
145  32.0          4          83.0        61.0    2003          19.0   
181  33.0          4          91.0        53.0    1795          17.5   
195  29.0          4          85.0        52.0    2035          22.2   
196  24.5          4          98.0        60.0    2164          22.1   
198  33.0          4          91.0        53.0    1795          17.4   
216  31.5          4          98.0        68.0    2045          18.5   
218  36.0          4          79.0        58.0    1825          18.6   
244  43.1          4          90.0        48.0    1985          21.5   
246  32.8          4          78.0        52.0    1985          19.4   
247  39.4          4          85.0        70.0    2070          18.6   
303  31.8          4          85.0        65.0    2020          19.2   
310  38.1          4          89.0        60.0    1968          18.8   
322  46.6          4          86.0        65.0    2110          17.9   
324  40.8          4          85.0        65.0    2110          19.2   
325  44.3          4          90.0        48.0    2085          21.7   
331  33.8          4          97.0        67.0    2145          18.0   
346  32.3          4          97.0        67.0    2065          17.8   
347  37.0          4          85.0        65.0    1975          19.4   
348  37.7          4          89.0        62.0    2050          17.3   
376  37.0          4          91.0        68.0    2025          18.2   
377  31.0          4          91.0        68.0    1970          17.6   
379  36.0          4          98.0        70.0    2125          17.3   
394  44.0          4          97.0        52.0    2130          24.6   

     model year  origin                         car name  
19           70       2     volkswagen 1131 deluxe sedan  
51           71       2                      peugeot 304  
53           71       3              toyota corolla 1200  
54           71       3                      datsun 1200  
55           71       2             volkswagen model 111  
56           71       1                 plymouth cricket  
79           72       2                  renault 12 (sw)  
102          73       2          volkswagen super beetle  
117          73       2                         fiat 128  
129          74       3                      datsun b210  
131          74       3              toyota corolla 1200  
145          74       3                       datsun 710  
181          75       3                 honda civic cvcc  
195          76       1               chevrolet chevette  
196          76       1                  chevrolet woody  
198          76       3                      honda civic  
216          77       3                honda accord cvcc  
218          77       2                    renault 5 gtl  
244          78       2  volkswagen rabbit custom diesel  
246          78       3                 mazda glc deluxe  
247          78       3                   datsun b210 gx  
303          79       3                       datsun 210  
310          80       3            toyota corolla tercel  
322          80       3                        mazda glc  
324       

### Masking Operation

The masking operation replaces values where the condition is True.

Consider the below table with student marks:
<table>
    <tr>
        <th>Students</th>
        <th>Chemistry</th>
        <th>Physics</th>
        <th>Mathematics</th>
        <th>English</th>
    </tr>
    <tr>
        <td>Subodh</td>
        <td>67</td>
        <td>45</td>
        <td>50</td>
        <td>19</td>
    </tr>
    <tr>
        <td>Ram</td>
        <td>90</td>
        <td>92</td>
        <td>87</td>
        <td>90</td>
    </tr>
    <tr>
        <td>Abdul</td>
        <td>66</td>
        <td>72</td>
        <td>81</td>
        <td>72</td>
    </tr>
    <tr>
        <td>John</td>
        <td>32</td>
        <td>40</td>
        <td>12</td>
        <td>68</td>
    </tr>
</table>

#### Problem Statement:
The teacher does not want to reveal the marks of students who have failed. The condition is that if a student has scored marks >= 33, then they have passed, otherwise failed. The marks of failed students has to be replaced with ‘Fail’. So, how can the task be performed?.

#### Solution:
First is to create the DataFrame as shown below:

In [48]:
marks = [{'Chemistry': 67, 'Physics': 45, 'Mathematics': 50, 'English' : 19},
        {'Chemistry': 90, 'Physics': 92, 'Mathematics': 87, 'English' : 90}, 
        {'Chemistry': 66, 'Physics': 72, 'Mathematics': 81, 'English' : 72}, 
        {'Chemistry': 32, 'Physics': 40, 'Mathematics': 12, 'English' : 68}]
marks_df = pd.DataFrame(marks, index = ['Subodh', 'Ram', 'Abdul', 'John'])
marks_df

Chemistry  Physics  Mathematics  English
Subodh         67       45           50       19
Ram            90       92           87       90
Abdul          66       72           81       72
John           32       40           12       68

<b>Syntax:</b>
<ul type='none' line-space="4px">
    <li><b>DataFrame.mask(cond, other = nan, inplace = False, axis = None)</b></li>
    <li>cond – Where cond is False, keep the original value. Where True, replace with corresponding value from other</li>
    <li>other - Entries where cond is True are replaced with corresponding value from other.</li>
    <li>inplace - Whether to perform the operation in place on the data.</li>
    <li>axis – alignment axis</li>
</ul>

In [49]:
f = marks_df < 33
marks_df.mask(f, 'Fail')

Chemistry  Physics Mathematics English
Subodh        67       45          50    Fail
Ram           90       92          87      90
Abdul         66       72          81      72
John        Fail       40        Fail      68

### Sorting Data

#### Problem Statement:
XYZ Custom cars want the data sorted according to the number of cylinders. 

#### Solution:
The following method can be used to get the solution.

In [50]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 392 entries, 0 to 397
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   mpg           392 non-null    float64
 1   cylinders     392 non-null    int64  
 2   displacement  392 non-null    float64
 3   horsepower    392 non-null    float64
 4   weight        392 non-null    int64  
 5   acceleration  392 non-null    float64
 6   model year    392 non-null    int64  
 7   origin        392 non-null    int64  
 8   car name      392 non-null    object 
dtypes: float64(4), int64(4), object(1)
memory usage: 38.7+ KB


In [53]:
df.sort_values(by = 'cylinders')

mpg  cylinders  displacement  horsepower  weight  acceleration  \
243  21.5          3          80.0       110.0    2720          13.5   
334  23.7          3          70.0       100.0    2420          12.5   
111  18.0          3          70.0        90.0    2124          13.5   
71   19.0          3          70.0        97.0    2330          13.5   
237  30.5          4          98.0        63.0    2051          17.0   
..    ...        ...           ...         ...     ...           ...   
86   14.0          8         304.0       150.0    3672          11.5   
85   13.0          8         350.0       175.0    4100          13.0   
285  17.0          8         305.0       130.0    3840          15.4   
92   13.0          8         351.0       158.0    4363          13.0   
0    18.0          8         307.0       130.0    3504          12.0   

     model year  origin                   car name  
243          77       3                 mazda rx-4  
334          80       3              mazda rx-7 gs  
111          73       3                  maxda rx3  
71           72       3            mazda rx2 coupe  
237          77       1         chevrolet chevette  
..          ...     ...                        ...  
86           73       1                amc matador  
85           73       1          buick century 350  
285          79       1  chevrolet caprice classic  
92           73       1                   ford ltd  
0            70       1  chevrolet chevelle malibu  

[392 rows x 9 columns]

#### Problem Statement:

There is a requirement in which the cars that have lowest acceleration must be assessed. It is also to be checked that which cars have higher horsepower despite having lower acceleration.

#### Solution:

In this case, the data must be sorted in ascending order of acceleration and descending order of horsepower as follows:

In [54]:
df.sort_values(['acceleration', 'horsepower'], ascending = (1,0))

mpg  cylinders  displacement  horsepower  weight  acceleration  \
11   14.0          8         340.0       160.0    3609           8.0   
7    14.0          8         440.0       215.0    4312           8.5   
9    15.0          8         390.0       190.0    3850           8.5   
6    14.0          8         454.0       220.0    4354           9.0   
116  16.0          8         400.0       230.0    4278           9.5   
..    ...        ...           ...         ...     ...           ...   
195  29.0          4          85.0        52.0    2035          22.2   
59   23.0          4          97.0        54.0    2254          23.5   
326  43.4          4          90.0        48.0    2335          23.7   
394  44.0          4          97.0        52.0    2130          24.6   
299  27.2          4         141.0        71.0    3190          24.8   

     model year  origin            car name  
11           70       1  plymouth 'cuda 340  
7            70       1   plymouth fury iii  
9            70       1  amc ambassador dpl  
6            70       1    chevrolet impala  
116          73       1  pontiac grand prix  
..          ...     ...                 ...  
195          76       1  chevrolet chevette  
59           72       2   volkswagen type 3  
326          80       2  vw dasher (diesel)  
394          82       2           vw pickup  
299          79       2         peugeot 504  

[392 rows x 9 columns]

### Preserving index

Pandas preserves the index and column labels in the output. For binary operations such as addition and multiplication, Pandas will automatically align indices when passing the objects to the functions. 

<table>
    <tr>
        <th>Students</th>
        <th>Chemistry</th>
        <th>Physics</th>
        <th>Mathematics</th>
        <th>English</th>
    </tr>
    <tr>
        <td>Subodh</td>
        <td>67</td>
        <td>45</td>
        <td>50</td>
        <td>19</td>
    </tr>
    <tr>
        <td>Ram</td>
        <td>90</td>
        <td>92</td>
        <td>87</td>
        <td>90</td>
    </tr>
    <tr>
        <td>Abdul</td>
        <td>66</td>
        <td>72</td>
        <td>81</td>
        <td>72</td>
    </tr>
    <tr>
        <td>John</td>
        <td>32</td>
        <td>40</td>
        <td>12</td>
        <td>68</td>
    </tr>
</table>


#### Problem Statement:
The teacher wants to encrypt the marks for confidential reasons. Therefore, the teacher decides to save the marks as sine of the original marks. For example, if Subodh has scored 67 in chemistry, then his encrypted marks will be sin(67) = -0.855520

#### Solution:
To get the solution, following code must be used:

In [55]:
marks = {'Chemistry': [67,90,66,32], 
        'Physics': [45,92,72,40],  
        'Mathematics': [50,87,81,12],  
        'English': [19,90,72,68]}
marks_df = pd.DataFrame(marks, index = ['Subodh', 'Ram', 'Abdul', 'John'])
marks_df

Chemistry  Physics  Mathematics  English
Subodh         67       45           50       19
Ram            90       92           87       90
Abdul          66       72           81       72
John           32       40           12       68

In [56]:
#encrypting marks as sine of marks
encrypted_marks = np.sin(marks_df)
encrypted_marks

Chemistry   Physics  Mathematics   English
Subodh  -0.855520  0.850904    -0.262375  0.149877
Ram      0.893997 -0.779466    -0.821818  0.893997
Abdul   -0.026551  0.253823    -0.629888  0.253823
John     0.551427  0.745113    -0.536573 -0.897928

The  encrypted marks are with same indices as the original marks. This is called as index preservation.

#### Resetting Index:

In case of a requirement where the index has to be restored to the default index, reset_index() function must be used. It adds the existing index as a new column in the DataFrame. This can be done as follows:

In [57]:
encrypted_marks.reset_index(inplace = True)
encrypted_marks

index  Chemistry   Physics  Mathematics   English
0  Subodh  -0.855520  0.850904    -0.262375  0.149877
1     Ram   0.893997 -0.779466    -0.821818  0.893997
2   Abdul  -0.026551  0.253823    -0.629888  0.253823
3    John   0.551427  0.745113    -0.536573 -0.897928

### Broadcasting operation

Broadcasting refers to a set of rules to operate between data of different sizes and shapes.

Consider the following table:

<table>
    <tr>
        <th>Students</th>
        <th>Chemistry</th>
        <th>Physics</th>
        <th>Mathematics</th>
        <th>English</th>
    </tr>
    <tr>
        <td>Subodh</td>
        <td>67</td>
        <td>45</td>
        <td>50</td>
        <td>19</td>
    </tr>
    <tr>
        <td>Ram</td>
        <td>90</td>
        <td>92</td>
        <td>87</td>
        <td>90</td>
    </tr>
    <tr>
        <td>Abdul</td>
        <td>66</td>
        <td>72</td>
        <td>81</td>
        <td>72</td>
    </tr>
    <tr>
        <td>John</td>
        <td>32</td>
        <td>40</td>
        <td>12</td>
        <td>68</td>
    </tr>
</table>


#### Problem Statement:
The teacher wants to award five bonus marks to all the students.

#### Solution:
This can be done by using broadcasting methods available in Pandas.

In [58]:
new_marks = marks_df + 5
new_marks

Chemistry  Physics  Mathematics  English
Subodh         72       50           55       24
Ram            95       97           92       95
Abdul          71       77           86       77
John           37       45           17       73

#### Problem Statement:
The teacher wants to increase the marks of all the students as follows-

Chemistry: + 5

Physics: + 10

Mathematics: +10

English: + 2

#### Solution:
This can be done as follows:

In [59]:
new_marks = marks_df + [5,10,10,2]
new_marks

Chemistry  Physics  Mathematics  English
Subodh         72       55           60       21
Ram            95      102           97       92
Abdul          71       82           91       74
John           37       50           22       70

#### Apply

This method is used to apply a function along an axis of the DataFrame.

Syntax:
<ul type="none">
       <li><b>DataFrame.apply(func, axis = 0, result_type = None)</b></li>
       <li>func : Function to apply to each column or row.</b></li>
       <li>axis: Axis along which the function is applied.</li>
       <li>result_type: one out of 'expand', 'reduce' or 'broadcast'. In the demo, 'broadcast' is used.</li>
        <li><ul>
            <li>‘broadcast’ : results will be broadcast to the original shape of the DataFrame, the original index and columns will be retained.</li>
        </ul></li>
</ul>

<table>
    <tr>
        <th>Students</th>
        <th>Chemistry</th>
        <th>Physics</th>
        <th>Mathematics</th>
        <th>English</th>
    </tr>
    <tr>
        <td>James</td>
        <td>67</td>
        <td>45</td>
        <td>50</td>
        <td>19</td>
    </tr>
    <tr>
        <td>Lee</td>
        <td>90</td>
        <td>92</td>
        <td>87</td>
        <td>90</td>
    </tr>
    <tr>
        <td>Anderson</td>
        <td>66</td>
        <td>72</td>
        <td>81</td>
        <td>72</td>
    </tr>
    <tr>
        <td>John</td>
        <td>32</td>
        <td>40</td>
        <td>12</td>
        <td>68</td>
    </tr>
</table>



In [61]:
#Creating the DataFrame
marks = {'Chemistry': [67,90,66,32], 
        'Physics': [45,92,72,40],  
        'Mathematics': [50,87,81,12],  
        'English': [19,90,72,68]}
marks_df = pd.DataFrame(marks, index = ['James', 'Lee', 'Anderson', 'John'])
marks_df

Chemistry  Physics  Mathematics  English
James            67       45           50       19
Lee              90       92           87       90
Anderson         66       72           81       72
John             32       40           12       68

#### Problem statement : 
The teacher wants to get the total marks scored in each subject

#### Solution :
This can be done as follows:

In [62]:
marks_df.apply(np.sum, axis = 0)

Chemistry      255
Physics        249
Mathematics    230
English        249
dtype: int64

#### Problem statement: 
The teacher wants to get the total marks scored by each student.

#### Solution:
This can be done as follows:

In [63]:
marks_df.apply(np.sum, axis = 1)

James       181
Lee         359
Anderson    291
John        152
dtype: int64

#### Problem Statement:
The students were unable to attend the next set of exams due to the pandemic. Hence, the teacher decides to award them average marks based on their previous performance.

#### Solution:
This can be done as follows:

In [64]:
marks_df.apply(func = np.mean, axis = 0, result_type = 'broadcast')

Chemistry  Physics  Mathematics  English
James            63       62           57       62
Lee              63       62           57       62
Anderson         63       62           57       62
John             63       62           57       62

### Aggregation operation in Pandas

#### Problem Statement:
Consider the scenario where the board of XYZ custom cars wants to know about minimum and maximum of all the numerical columns.

#### Solution:
Aggregation operation is used to aggregate using one or more operations over the specified axis.

<b>Syntax:

DataFrame.agg(func, axis = 0)</b>

func - Function to use for aggregating the data. If a function, must either work when passed a DataFrame or when  passed to DataFrame.apply.

axis: If 0 or ‘index’: apply function to each column. If 1 or ‘columns’: apply function to each row.

The below code is used to find minimum and maximum values of the numerical attributes:

In [65]:
#Using list comprehension to get the numerical columns
list1 = [col for col in df.columns if df[col].dtype in ['float', 'int64']]
df[list1].agg(['min', 'max'])

mpg  cylinders  displacement  horsepower  weight  acceleration  \
min   9.0          3          68.0        46.0    1613           8.0   
max  46.6          8         455.0       230.0    5140          24.8   

     model year  origin  
min          70       1  
max          82       3

### Grouping Operation in Pandas


XYZ custom cars want to know the number of cars manufactured in each year.

This would require a grouping operation. Pandas supports a group by feature to group our data for aggregate operations.

Syntax:

       DataFrame.groupby(by = column_name, axis, sort)

#### Problem statement:
How many cars belong to each year?

#### Solution:
In addition to the groupby function, the count function can be used as shown below. Since, cars are counted by names in each model year, the ‘name’ column in a list is used to get the output as a DataFrame.

In [68]:
df.groupby(['model year']).count()[['car name']]

car name
model year          
70                29
71                27
72                28
73                40
74                26
75                30
76                34
77                28
78                36
79                29
80                27
81                28
82                30

#### Problem Statement: 
Some senior engineers in XYZ custom cars want to understand about the effect of model year and number of cylinders on horsepower.

#### Solution: 
One of the engineers suggests about checking the mean, minimum and maximum horsepower based on number of cylinders and model year. For such requirement, the ‘agg’ function can be combined with groupby function as shown below:

In [69]:
#Creating a DataFrame grouped on cylinders and model_year and finding mean, min and max of horsepower
grouped_multiple = df.groupby(['cylinders', 'model year']).agg({'horsepower': ['mean', 'min', 'max']})
#Naming columns in grouped DataFrame
grouped_multiple.columns = ['hp_mean', 'hp_min', 'hp_max']
#Resetting index
grouped_multiple = grouped_multiple.reset_index()
#Viewing head of resulting DataFrame
grouped_multiple.head()

cylinders  model year     hp_mean  hp_min  hp_max
0          3          72   97.000000    97.0    97.0
1          3          73   90.000000    90.0    90.0
2          3          77  110.000000   110.0   110.0
3          3          80  100.000000   100.0   100.0
4          4          70   87.714286    46.0   113.0

#### Problem statement: 
The engineers at XYZ Custom Cars want to know about the relationship between model year and acceleration of cars.

#### Solution:
For better understanding, the grouped results can be sorted based on average acceleration of cars built in each model year.

In [80]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 392 entries, 0 to 397
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   mpg           392 non-null    float64
 1   cylinders     392 non-null    int64  
 2   displacement  392 non-null    float64
 3   horsepower    392 non-null    float64
 4   weight        392 non-null    int64  
 5   acceleration  392 non-null    float64
 6   model year    392 non-null    int64  
 7   origin        392 non-null    int64  
 8   car name      392 non-null    object 
dtypes: float64(4), int64(4), object(1)
memory usage: 38.7+ KB


In [81]:

df.groupby(['model year']).sort_values('acceleration',ascending = False)[['acceleration']]


# df.groupby(['model year']).mean()#.sort_values('acceleration', ascending = False)[['acceleration']]

AttributeError: 'DataFrameGroupBy' object has no attribute 'sort_values'